src/can.rs

```diff
@@ -0,0 +1,111 @@
+//! Controller Area Network
+
+use core::convert::TryInto;
+use core::convert::TryFrom;
+
+use nb;
+
+/// A type that can either be `BaseId` or `ExtendedId`
+#[cfg(feature = "unproven")]
+pub trait Id: Sized {
+    /// The BaseId variant
+    type BaseId: BaseId + Into<Self> + TryFrom<Self>;
+
+    /// The ExtendedId variant
+    type ExtendedId: ExtendedId + Into<Self> + TryFrom<Self>;
+}
+
+/// A Can (11-bit) ID
+#[cfg(feature = "unproven")]
+pub trait BaseId: Sized {
+    /// A generic ID type that encapsulate this type
+    type Id: Id + From<Self> + TryInto<Self>;
+}
+
+/// A Can Extended (28-bit) ID
+#[cfg(feature = "unproven")]
+pub trait ExtendedId: Sized {
+    /// A generic ID type that encapsulate this type
+    type Id: Id + From<Self> + TryInto<Self>;
+}
+
+/// A Can Frame
+#[cfg(feature = "unproven")]
+pub trait Frame {
+    /// The Id type of this Frame
+    type Id: Id;
+
+    /// Returns the Can-ID
+    fn id(&self) -> Self::Id;
+
+    /// Returns `Some(Data)` if data frame.
+    /// Returns `None` if remote frame.
+    fn data(&self) -> Option<&[u8]>;
+}
+
+/// A Can-FD Frame
+///
+/// A "ordinary" Can-Frame must also be representable by this type.
+#[cfg(feature = "unproven")]
+pub trait FdFrame {
+    /// The Id type of this Frame
+    type Id: Id;
+
+    /// Returns true if this frame would/has be(en) transmitted as a Can-Fd frame.
+    /// Returns false if this frame would/has be(en) transmitted as a "ordinary" Can frame.
+    fn is_fd_frame(&self) -> bool;
+
+    /// Returns the Can-ID
+    fn id(&self) -> Self::Id;
+
+    /// Returns `Some(Data)` if data frame.
+    /// Returns `None` if remote frame.
+    fn data(&self) -> Option<&[u8]>;
+}
+
+
+/// A CAN interface
+#[cfg(feature = "unproven")]
+pub trait Interface {
+    /// The Can Frame this Interface operates on
+    type Frame: Frame;
```

Comment:
Doesn't this associated type mean that an interface only works with one frame type?  If so, and the frame type is associated with the Id type, you'd need two different interfaces for the same hardware in the event of a bus that contains both id types.  That doesn't strike me as a good design.

Reply:
In Frame2, what would be the type of the id?

Reply:
Probably something close to:
``` rust
pub struct Id2 {
    extended: bool,
    id: u32,
}
```